Add API to call callbacks after a cancel

paulb777 opened the pull request:
Fix #318

mwyman commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -139,6 +139,7 @@ @interface GTMSessionFetcher ()
 
 @property(atomic, strong, readwrite, nullable) NSData *downloadedData;
 @property(atomic, strong, readwrite, nullable) NSData *downloadResumeData;
+@property(atomic, readwrite) BOOL cancelAndCallback;
```
Note: all the fetcher's `atomic` properties are explicitly implemented and do `@synchronized(self)` rather than rely on the generated "atomic" getter/setter (which almost never does what devs really want).

mwyman commented on Sources/Core/Public/GTMSessionFetcher/GTMSessionFetcher.h:
```diff
@@ -1103,6 +1104,10 @@ __deprecated_msg("implement GTMSessionFetcherAuthorizer instead")
 
+// Cancel the fetch of the request that's currently in progress.  The completion handler
+// is called on option.
+- (void)stopFetching:(BOOL)shouldCallCallbacks;
```
My understanding is that this change is being driven by desire for Swift `async` support, and if Swift code demands the continuation is always called then the API should make sure than an appropriately Swift-configured fetcher will always trigger callbacks on `stopFetching` being called, without the code canceling the fetch needing to know to call a different `stopFetching` variant.

I think this needs to be reconsidered and made into a single fire-and-forget kind of approach, where Swift code can set a `fetcher.stopFetchingTriggersCallbacks = true` kind of property, at which point any call to `stopFetching` will trigger the callbacks with the appropriate error.

mwyman commented on UnitTests/GTMSessionFetcherFetchingTest.m:
```diff
@@ -1720,6 +1720,40 @@ - (void)testQuickBeginStopFetching {
 #endif
 }
 
+- (void)testQuickBeginDontStopFetching {
```
This seems unrelated to the PR at hand; can we simply handle the completion callback-with-error behavior, and assert that cancellation triggers an appropriate error in the fetcher?

Consider having your test access the local HTTP `_testServer` with a "sleep=" query parameter so you can have time to cancel the fetch on the client side.

paulb777 replied:
This test directly contrasts with the test before it to verify that the callbacks do get called, where they didn't in the previous test. Isn't that useful?

mwyman replied:
The test before this is testing a specific workaround; testing that the callback is triggered on cancel requires different behavior, which I described in my original comment.

mwyman commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -1964,14 +1966,26 @@ - (void)forgetSessionIdentifierForFetcherWithoutSyncCheck {
+    NSError *error = [NSError errorWithDomain:kGTMSessionFetcherErrorDomain
+                                   code:GTMSessionFetcherErrorCancelled
+                                     userInfo:@{@"description" : @"Operation cancelled"}];
+    _cancelAndCallback = YES;
```
Note that access to `atomic` properties underlying ivars are not implicitly synchronized (ever), and so either need to use `.` property access or be wrapped in the appropriate locking (if the getter/setter are implemented manually).

mwyman commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -1964,14 +1966,26 @@ - (void)forgetSessionIdentifierForFetcherWithoutSyncCheck {
+                                   code:GTMSessionFetcherErrorCancelled
+                                     userInfo:@{@"description" : @"Operation cancelled"}];
+    _cancelAndCallback = YES;
+    _completionHandler(nil, error);
```
The completion handler block should not be invoked directly from the thread calling `stopFetching`, among other things because the completion handler block must be invoked on the `callbackQueue`.

Also, access to the `_completionHandler` ivar needs to be guarded by `@synchronized(self)`.

It would probably be best to add parameter(s) to `stopFetchingReleasingCallbacks:` that triggers invoking the completion handler prior to calling `releaseCallbacks`, following the behavior pattern of `invokeFetchCallbacksOnCallbackQueueWithData:error:mayDecorate:shouldReleaseCallbacks:`, or possibly calls that method directly.

paulb777 reviewed (commented):
Thanks for the feedback. After being out much of the last two weeks, I'm finally back at it ...

mwyman commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -1972,7 +1972,19 @@ - (void)stopFetching {
   }  // @synchronized(self)
-  [self stopFetchReleasingCallbacks:YES];
+
+  if (self.stopFetchingTriggersCallbacks) {
+    NSError *error = [NSError errorWithDomain:kGTMSessionFetcherErrorDomain
+                                         code:GTMSessionFetcherErrorCancelled
+                                     userInfo:@{@"description" : @"Operation cancelled"}];
+    [self invokeFetchCallbacksOnCallbackQueueWithData:nil
+                                                error:error
+                                          mayDecorate:YES
+                               shouldReleaseCallbacks:YES];
+    [self sendStopNotificationIfNeeded];
```
This doesn't trigger stopping the fetch, just invokes the callback and sends the notification.

I think the changes needs to be spread across `stopFetchingReleasingCallbacks:` and `URLSession:task:didCompleteWithError:`.

Here, instead of this code we pass `NO` if `self.stopFetchingTriggersCallbacks` to this initial `stopFetchingReleasingCallbacks:` and in `URLSession:task:didCompleteWithError:` checks for an error with `NSURLErrorDomain` and code `NSURLErrorCancelled` and perhaps wraps that error before calling the callback.

WDYT?

paulb777 replied:
I couldn't get the changes to work in `stopFetchingReleasingCallbacks` because of other call paths into it.   Instead, I reverted to always calling  `stopFetchingReleasingCallbacks`  from `stopFetching`. 

I'm also not sure what to do in `URLSession:task:didCompleteWithError:`. Based on https://github.com/google/gtm-session-fetcher/blob/main/UnitTests/GTMSessionFetcherFetchingTest.m#L886, it looks like callbacks occur if the cancellation happens on the server.

mwyman replied:
My worry is that this is a new codepath to firing the completion handler, and I'm not sure the implications of firing the handler here before the rest of the flow has completed.

My suggestion is that we allow the task cancellation to take place, but in `finishWithError:shouldRetry:` we would want to check for a cancellation error from NSURLSession, and if so check whether `_userStoppedFetching` is set, and let the callback flow happen that way. Would also need to ensure the completion handler block will fire after user-stopped for that case.

When `cancel` is called on a task, `NSURLErrorCancelled` is returned: https://developer.apple.com/documentation/foundation/nsurlsessiontask/1411591-cancel

mwyman commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -2600,7 +2611,7 @@ - (void)invokeOnCallbackQueue:(dispatch_queue_t)callbackQueue
   if (callbackQueue) {
     dispatch_group_async(_callbackGroup, callbackQueue, ^{
-      if (!afterStopped) {
+      if (!afterStopped && !self->_stopFetchingTriggersCallbacks) {
```
I worry this will let other blocks slip through if `_stopFetchingTriggersCallbacks` is true, and that may not be what we want.

For Swift `async` my understanding is the only callback we want firing after `stopFetching` is called is the completion handler because that's the one that needs to trigger the continuation. Are we sure we want to allow other callback blocks to be sent? Otherwise, I think the `afterStopped` should be what we adjust if `.stopFetchingTriggersCallbacks` is set.

paulb777 replied:
I couldn't find a way to get the afterStopped parameter right. I did update the added test to limit to when triggersCompletion is on and there is a userStop occurrence.

thomasvl replied:
Does this still need work or did this get figured out?

paulb777 replied:
I've updated to limit the fall through to only when there has been a stop, so yes.

mwyman commented on Sources/Core/Public/GTMSessionFetcher/GTMSessionFetcher.h:
```diff
@@ -1100,9 +1101,12 @@ __deprecated_msg("implement GTMSessionFetcherAuthorizer instead")
 - (void)stopFetching;
 
+// Call callbacks in `stopFetching`.
+@property(atomic, assign) BOOL stopFetchingTriggersCallbacks;
```
Should this be limited to the completion handler?

`stopFetchingTriggersCompletionHandler` ?

paulb777 commented:
Moving to `finishWithError:shouldRetry:` seems to work except for the case where the cancel happens after a successful upload but before the success completion is called. I can't find where the continuation gets leaked in that case.

paulb777 commented:
Mostly addressed the comments. Open issues are three test failures because of fetcher.completionHander not being nilled out and correctly passing the `afterStopped` parameter

mwyman commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -2991,6 +3000,14 @@ - (void)URLSession:(NSURLSession *)session
+  if (self.stopFetchingTriggersCallbacks && self.userStoppedFetching) {
+    NSError *error = [NSError errorWithDomain:kGTMSessionFetcherErrorDomain
+                                         code:GTMSessionFetcherErrorCancelled
+                                     userInfo:@{@"description" : @"Operation cancelled"}];
```
@"description" -> NSLocalizedDescriptionKey

mwyman replied:
Would we want to also add NSUnderlyingErrorKey with the cancellation error from NSURLSession? Not sure whether I think so, but wanted to ask the question.

mwyman commented on UnitTests/GTMSessionFetcherChunkedUploadTest.m:
```diff
@@ -1537,6 +1577,63 @@ - (void)testBigDataChunkedUploadWithCancel {
+
+  NSURLRequest *request = [self validUploadFileRequest];
+  NSData *bigData = [self bigUploadData];
+  GTMSessionUploadFetcher *fetcher = [GTMSessionUploadFetcher uploadFetcherWithRequest:request
```
Mind making sure you've run `git clang-format` on this PR?

paulb777 commented:
The latest test should address the review comments and pass the tests. I still have some issues with Firebase tests to investigate.

paulb777 commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -2667,25 +2671,28 @@ - (void)invokeFetchCallbacksOnCallbackQueueWithData:(nullable NSData *)data
   }
 
   if (handler) {
-    [self invokeOnCallbackQueue:callbackQueue afterUserStopped:NO block:^{
```
clang-format change only

thomasvl commented:
I also did a clang-format on `main`, so if you rebase/merge, the diffs specific to that should go away.

paulb777 commented:
Rebased with no conflicts

paulb777 commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -1995,6 +1995,7 @@ - (void)stopFetchReleasingCallbacks:(BOOL)shouldReleaseCallbacks {
   [holdSelf destroyRetryTimer];
```
Updated

thomasvl replied:
Sorry, meant line 1981, it currently says:

```
// If shouldReleaseCallbacks is NO then the fetch will be retried so the callbacks
// need to still be retained.
```

Which isn't completely true any more, as the cancel behavior now factors in to when `NO` is used.

paulb777 replied:
Updated 1981.

paulb777 commented on Sources/Core/Public/GTMSessionFetcher/GTMSessionFetcher.h:
```diff
@@ -1084,9 +1085,12 @@ __deprecated_msg("implement GTMSessionFetcherAuthorizer instead")
 
 // Cancel the fetch of the request that's currently in progress.  The completion handler
-// will not be called.
+// will be called if the property `stopFetchingTriggersCompletionHandler` is `YES`.
```
Perhaps, it should only be supported on the service, since it doesn't really make sense to change once the fetcher has been created. WDYT?

thomasvl replied:
Actually, that might simplify things since it locks in the state so you don't need the synchronization, etc. Would that work for the Firebase usage?

paulb777 replied:
Actually no - since we want to be able to call a combination of new and old APIs.  However, it does seem reasonable to make `stopFetchingTriggersCompletionHandler` a readonly property and I did that in the latest commit.

paulb777 commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -2598,9 +2602,11 @@ - (void)invokeOnCallbackUnsynchronizedQueueAfterUserStopped:(BOOL)afterStopped
     dispatch_group_async(_callbackGroup, callbackQueue, ^{
-      if (!afterStopped) {
+      if (!afterStopped &&
+          (!self->_stopFetchingTriggersCompletionHandler || !self->_userStoppedFetching)) {
```
Added sync

paulb777 commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -2992,8 +2999,20 @@ - (void)URLSession:(NSURLSession *)session
+                                               code:GTMSessionFetcherErrorUserCancelled
+                                           userInfo:@{
+                                             NSLocalizedDescriptionKey : @"Operation cancelled",
+                                             NSUnderlyingErrorKey : error.localizedDescription
```
Fixed

thomasvl commented on Sources/Core/Public/GTMSessionFetcher/GTMSessionFetcherService.h:
```diff
@@ -136,9 +136,12 @@ extern NSString *const kGTMSessionFetcherServiceSessionKey;
 - (id)fetcherWithRequest:(NSURLRequest *)request fetcherClass:(Class)fetcherClass;
+- (id)fetcherWithRequest:(NSURLRequest *)request
+              fetcherClass:(Class)fetcherClass
+    stopTriggersCompletion:(BOOL)stopTriggersCompletion;
```
I was thinking we just make this a property on the service and then you don't need a new api, it just gets copied out like all the other properties that get carried over?

paulb777 replied:
That doesn't work if you want to use the service to call both sets of APIs. I suppose I could change Firebase to add a second service just for usage from async APIs?

thomasvl replied:
> Done.

Just checking, did that mean you had another commit to also push?  I don't think one appeared after this comment.

paulb777 replied:
Yep. I didn't notice that I typo'd the `git push` command. Thanks.

thomasvl commented on Sources/Core/Public/GTMSessionFetcher/GTMSessionFetcher.h:
```diff
@@ -1084,9 +1085,12 @@ __deprecated_msg("implement GTMSessionFetcherAuthorizer instead")
+// will be called if the property `stopFetchingTriggersCompletionHandler` is `YES`.
 - (void)stopFetching;
 
+// Call callbacks in `stopFetching`.
```
Document that it can't be changed after begin

paulb777 replied:
Done.

thomasvl commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -3995,6 +4019,21 @@ - (void)setUsingBackgroundSession:(BOOL)flag {
+        _initialBeginFetchDate == nil,
+        @"stopFetchingTriggersCompletionHandler should not change after fetcher starts");
+
+    _stopFetchingTriggersCompletionHandler = flag;
```
This should probably still have a guard so it won't update if _initialBeginFetchDate is non nil.

thomasvl replied:
Something like:

```
  if (_initialBeginFetchDate == nil) {
    _stopFetchingTriggersCompletionHandler = flag;
  } else {
    GTMSESSION_ASSERT_DEBUG(
        0,
        @"stopFetchingTriggersCompletionHandler should not change after fetcher starts");
  }
```

?

paulb777 replied:
Updated

thomasvl commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -3995,6 +4019,21 @@ - (void)setUsingBackgroundSession:(BOOL)flag {
+}
+
+- (void)setStopFetchingTriggersCompletionHandler:(BOOL)flag {
+  @synchronized(self) {
```
I think you can skip the sync, _initialBeginFetchDate gets set without an sync, so you're probably safe skipping the sync here also. If someone is setting up a fetcher from multiple threads, or starting it while setting it up from multiple threads, they already are in a risky space.

paulb777 replied:
done

thomasvl commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -2598,9 +2603,14 @@ - (void)invokeOnCallbackUnsynchronizedQueueAfterUserStopped:(BOOL)afterStopped
     dispatch_group_async(_callbackGroup, callbackQueue, ^{
-      if (!afterStopped) {
+      BOOL skipEarlyExit = NO;
+      @synchronized(self) {
```
this sync isn't needed now that it can't change after begin.

paulb777 replied:
removed

paulb777 reviewed (commented):
Thanks for the reviews!

thomasvl commented on Sources/Core/GTMSessionFetcher.m:
```diff
@@ -2066,6 +2068,7 @@ - (void)stopFetchReleasingCallbacks:(BOOL)shouldReleaseCallbacks {
         }
       }
     }
+    shouldStopFetcher = !_stopFetchingTriggersCompletionHandler || !_userStoppedFetching;
```
Why is this one `!_userStoppedFetcher`?

If `stopFetchingTriggersCompletionHandler` is `NO`, and `stopFetching` is called, then `[service fetcherDidStop:self]` on line 2090 will be skipped, so what actually does that work now?

At first reread, should this be `!_stopFetchingTriggersCompletionHandler && _userStoppedFetching`?  That would make the code continue the current behavior before adding `_stopFetchingTriggersCompletionHandler`.

paulb777 replied:
This change hangs the `testFetcherService` and other tests. I'll investigate.

paulb777 replied:
After further investigation, I believe the logic was correct. I inverted the condition and it's hopefully easier to understand.

thomasvl replied:
Thanks, yea, this does read easier.

thomasvl commented on Sources/Core/Public/GTMSessionFetcher/GTMSessionFetcher.h:
```diff
@@ -1084,9 +1085,12 @@ __deprecated_msg("implement GTMSessionFetcherAuthorizer instead")
 - (void)stopFetching;
 
+// Call callbacks in `stopFetching`. It cannot be changed once the fetcher starts.
+@property(atomic, assign) BOOL stopFetchingTriggersCompletionHandler;
```
Sorry, didn't think of this before – Can you add a block here speaking to the point if the fetcher is going to be use from Swift `async` calls, they this must be set to YES to ensure a cancel doesn't leave the Swift runtime in a bad state?.

paulb777 replied:
Done
